### PR TITLE
Fix #58, block compression incompatibility with python library

### LINKED ISSUE #58: Block compression incompatibility with C/Python library
Since commit https://github.com/pierrec/lz4/commit/a207029655a0aa70c4cba8263c5aa8d273f9b28c (identified through bisecting) compressing some data causes the Python/C decompressor to fail with error message `_block.LZ4BlockError: Decompression failed: corrupt input or insufficient space in destination buffer. Error code: 179`. 

The problem seems to be that the decompressor cannot fit the decompressed data within the provided buffer size (which is the size of the original data). Or is somehow tricked into thinking it cannot fit the data. Increasing the provided size by one resolves the problem in all cases observed so far.

Below is a Go compressor and Python decompressor pair that demonstrates the problem. The input data provided is the smallest I managed to identify that would reproduce the problem:

```go
package main

import (
	"fmt"
	"github.com/pierrec/lz4"   // v2.3.0
	"os"
)


func check(err error) {
	if err != nil {
		panic(err)
	}
}

func main() {
        // 214 bytes
	originalInput := []byte(`[{"rqgulitaeas_bitmap":0,"mwsswng_afnce":null,"fm_rxposire_op":null,"lnpmct_abs_diff_pct":null,"pqrstuvwyx_pxtch_eating":null,"fm_wxpssxre_rbs_dwxx_pct":null,"ued_date":"2020-09-15","regglttabns_bitmap_op":0,"dfff_`)
	input := originalInput
	var ht [1 << 16]int
	dst := make([]byte, lz4.CompressBlockBound(len(input)))
	fmt.Printf("Dst size: %d, input size: %d, original size: %d\n", len(dst), len(input), len(originalInput))
       
	size, err := lz4.CompressBlock(input, dst, ht[:])
	check(err)

	if size == 0 {
		fmt.Println("Uncompressible!")
		return
	}

        // No problem decompressing using Go lz4 with the exact size of the input
	uncompressedBuf := make([]byte, len(input))
	_, err = lz4.UncompressBlock(dst[:size], uncompressedBuf)
	check(err)

	f, err := os.Create(fmt.Sprintf("test_compressed_%d.lz4", len(input)))
	check(err)
	defer f.Close()

	_, err = f.Write(dst[:size])
	check(err)
}
```

```python
# Python 3.7.4
# LZ4 2.2.1

import lz4.block

size = 214
with open(f'test_compressed_{size}.lz4', mode='rb') as f:
    # This crashes
    lz4.block.decompress(f.read(), uncompressed_size=size)
   
    # This works
    # lz4.block.decompress(f.read(), uncompressed_size=size+1)
```

Is this a bug in github.com/pierrec/lz4, a bug in the C implementation or something I just have to live with and deal with?

### PR DESCRIPTION
**Pull request by tobgu**
Did the minimal update suggested in #58 which solved the problem (thanks @klauspost )! No new tests added at this stage. Suggestions for relevant tests (without pulling in the python library) are welcome if such exist and you think they should be added.

### CODE REVIEW

**Comment by klauspost** [comment]
The reason for this change is that the match search looks for matches from `si` to `si+2` in each loop. Therefore we could be within the apparent 12 byte end window.

The [match extension](https://github.com/pierrec/lz4/blob/master/block.go#L127) could be changed to be `for si <= sn `, but it would only rarely make any difference. I guess it could actually be `for sn-si >= 8 {`

**Comment by pierrec** [comment]
Gotcha. Thanks for the PR and explanation.
@tobgu I think that what @klauspost mentions in the second part of his explanation is the way to go.
Could you please amend your PR accordingly?
If you can't add a test for it, I will.

**Comment by tobgu** [comment]
Thanks for your input!

I basically code in the dark here based on you comments as I haven't had the time to actually dig into what the code is doing or lz4 internals in general.

Reverting the previous change and changing the match extension from `for si < sn {` -> `for sn-si >= 8 {` increases the output size from 190 to 194 bytes but the Python lib still crashes with the same error as before.

Reverting the previous change and changing the match extension from `for si < sn {` -> `for si < sn {` makes no difference at all for this particular case.

Keeping the previous change and updating the match extension to `for sn-si >= 8 {` works. I've updated the PR with that.

File: block.go [hunk 1]
```diff
@@ -124,7 +124,7 @@ func CompressBlock(src, dst []byte, hashTable []int) (_ int, err error) {
 		si, mLen = si+mLen, si+minMatch
 
 		// Find the longest match by looking by batches of 8 bytes.
-		for si < sn {
+		for sn-si >= 8 {
```
Comment (pierrec) [inline block.go]: Should actually be:
for si+8 < sn {

**Comment by pierrec** [comment]
I still cannot reproduce with pure go lib. Try the change in my comment on your PR. The check was actually incorrect.

File: block.go (hunk 1 above)
Comment (tobgu) [reply block.go]: Sorry for the long delay, holidays and all...

I've updated the condition. It does not seem to have any significant impact in this case. 

Did you see the comment in https://github.com/python-lz4/python-lz4/issues/183? Do you know how it compares to compressing it with the C-library directly?

**Pull request merged**